pkg/controller/autoscaling/elasticsearch/controller.go

```diff
@@ -187,7 +188,16 @@ func newElasticsearchClient(
 	es esv1.Elasticsearch,
 ) (esclient.Client, error) {
 	defer tracing.Span(&ctx)()
-	url := services.ExternalServiceURL(es)
+	resourcesState, err := esreconciler.NewResourcesStateFromAPI(c, es)
+	if err != nil {
+		return nil, err
+	}
+	url, err := services.ElasticsearchURLFromRandomPod(es, resourcesState.CurrentPodsByPhase[corev1.PodRunning])
+	if err != nil {
+		log := logconf.FromContext(ctx)
+		log.V(1).Info("could not get a URL to communicate with elasticsearch from a random pod, failing back to using external service", "error", err)
+		url = services.ElasticsearchURL(es, resourcesState.CurrentPodsByPhase[corev1.PodRunning])
```

Comment:
Same comment as the one on `driver.go`. Also it looks like duplicated code, which might mean that we want the same behaviour in all the cases:
* Try to get the URL from a Pod as a best effort
* Fallback on the external service if it is not possible

Should we just implement this mechanism in `ElasticsearchURLFromRandomPod` ?

Reply:
This is how I initially had things setup, but upon some comments from @pebrc  https://github.com/elastic/cloud-on-k8s/pull/5005#discussion_r745055036, I updated it to do just one thing, and error if it can't.  

Do both of you prefer for this func to just fail back if it can't generate the URL inside this func?

Reply:
My comment was more about the naming of the function. Not so much about not being able to have a shared function with a fallback. 

I wonder why we are not simply changing the existing `services.ElasticsearchURL` function to have the new behaviour of picking a random Pod by default and falling back to the service URL if that fails for some reason. 

Because the flow we have right now in this PR seems a bit odd: 

We call first `services.ElasticsearchURLFromRandomPod` if that fails we call `services.ElasticsearchURL` which in case of a scheme change once more calls `services.ElasticsearchURLFromRandomPod` (which we already know that it fails) 

I believe if we default to picking a random pod we do not need the scheme change logic at all anymore and we can keep the service URL as a fallback.